### PR TITLE
Add gaussian noise transform #6192

### LINKED ISSUE #6192: Add Gaussian noise transformation
### 🚀 The feature

Add gaussian noise transformation in the functionalities of torchvision.transforms.

### Motivation, pitch

Using Normalizing Flows, is good to add some light noise in the inputs. Right now I am using albumentation for this but, would be great to use it in the torchvision library

### Alternatives

[Albumentation has a gaussian noise implementation](https://albumentations.ai/docs/api_reference/augmentations/transforms/#albumentations.augmentations.transforms.GaussianBlur)

### Additional context

_No response_

cc @vfdev-5 @datumbox

### PR DESCRIPTION
**Pull request by parth-shastri**
Fixes #6192

Adds GaussianNoise transform functionality to the torchvision.transform.

-Basic GaussianNoise augmentation technique

### CODE REVIEW

**Comment by parth-shastri** [comment]
This PR addresses the issue #6192.

File: torchvision/transforms/transforms.py [hunk 1]
```diff
@@ -1837,6 +1838,80 @@ def __repr__(self) -> str:
...
+            sampling the gaussian noise. If float, sigma is fixed. If it is tuple
+            of float (min, max), sigma is chosen uniformly at random to lie in the
+            given range.
+
+    Returns:
+        PIL Image or Tensor: Gaussian blurred version of the input image.
+
+    """
+
+    def __init__(self, mean, sigma=(0.1, 2.0)):
```
Comment (oke-aditya) [inline torchvision/transforms/transforms.py:1858]:
```suggestion
    def __init__(self, mean: float, sigma: Optional[float, Tuple[float, float]] = (0.1, 2.0)):
```

Not sure if Tuple[float, float] goes well with with JIT. Might need to check.

File: torchvision/transforms/transforms.py [hunk 2]
```diff
@@ -1837,6 +1838,80 @@ def __repr__(self) -> str:
...
+
+    def forward(self, image: Tensor) -> Tensor:
+        """
+        Args:
+            image (PIL Image or Tensor): image to be perturbed with gaussian noise.
+
+        Returns:
+            PIL Image or Tensor: Image added with gaussian noise.
+        """
+        if not isinstance(image, torch.Tensor):
```
Comment (oke-aditya) [inline torchvision/transforms/transforms.py:1890]: I'm bit hesitant about this implementation. Why not creation `F.gaussian_noise` that would work well on PIL Images and tensors which we can then use here?

See https://github.com/pytorch/vision/blob/main/torchvision/transforms/functional.py#L1338

File: test/test_transforms.py [hunk 3]
```diff
@@ -1314,6 +1314,25 @@ def test_gaussian_blur_asserts():
         transforms.GaussianBlur(3, "sigma_string")
 
 
+def test_gaussian_noise():
+    np_img = np.ones((100, 100, 3), dtype=np.uint8) * 255
+    img = F.to_pil_image(np_img, "RGB")
+    out = transforms.GaussianNoise(2.0, (0.1, 2.0))(img)
+    assert F._is_pil_image(out) is True
```
Comment (oke-aditya) [inline test/test_transforms.py:1321]: I don't think so we need to assert this as this is independent job of `F._is_pil_image` which should just work.

File: torchvision/transforms/transforms.py (hunk 2 above)
Comment (oke-aditya) [inline torchvision/transforms/transforms.py:1890]: Probably sigma is not initialized. You are missing this?

```suggestion
        sigma = self.get_params(self.sigma[0], self.sigma[1])
        if not isinstance(image, torch.Tensor):
```

File: torchvision/transforms/transforms.py [hunk 4]
```diff
@@ -1837,6 +1838,80 @@ def __repr__(self) -> str:
...
+
+    Args:
+        mean (float or sequence): Mean of the sampling gaussian distribution .
+        sigma (float or tuple of float (min, max)): Standard deviation to be used for
+            sampling the gaussian noise. If float, sigma is fixed. If it is tuple
+            of float (min, max), sigma is chosen uniformly at random to lie in the
+            given range.
+
+    Returns:
+        PIL Image or Tensor: Gaussian blurred version of the input image.
```
Comment (oke-aditya) [inline torchvision/transforms/transforms.py:1854]: Hmm maybe perhaps we can say
```suggestion
        PIL Image or Tensor: Input image perturbed with Gaussian Noise.
```

File: torchvision/transforms/transforms.py [hunk 5]
```diff
@@ -46,6 +46,7 @@
     "RandomPerspective",
     "RandomErasing",
     "GaussianBlur",
+    "GaussianNoise",
```
Comment (oke-aditya) [inline torchvision/transforms/transforms.py:49]: You also need to add into docs. 

Edit 

https://github.com/pytorch/vision/blob/main/docs/source/transforms.rst  and

https://github.com/pytorch/vision/blob/main/gallery/plot_transforms.py#L121

**Review by oke-aditya — COMMENTED** [review]
Hi @parth-shastri thanks a lot for the PR.

While I haven't gone through thorough details and verified the implementation fully. I had few code style and comments.

Also, 

[torchvision/_C.pyd](https://github.com/pytorch/vision/pull/6233/files#diff-b1c3e3207017c0e86c8f70a07af1a1d3d1f356c6f9578021254b30fd51eb22cb)

file is unnecessary. Is it because you are using windows and this file is generated while compiling torchvision from source?

I am sure @vfdev-5 is champion in this domain and will have a deep look :smile:

File: torchvision/transforms/transforms.py (hunk 2 above)
Comment (vfdev-5) [reply torchvision/transforms/transforms.py:1890]: I agree, let's create a functional op `F.gaussian_noise` and at first iteration we can have tensor implementation. If input is PIL, we can convert it to tensor, apply the op and get back PIL image type for the output.

File: torchvision/transforms/transforms.py (hunk 5 above)
Comment (parth-shastri) [reply torchvision/transforms/transforms.py:49]: Yes will do.

File: torchvision/transforms/transforms.py (hunk 1 above)
Comment (oke-aditya) [reply torchvision/transforms/transforms.py:1858]: Aah yes my bad

**Comment by vfdev-5** [comment]
@parth-shastri can you please fix failing job "ci/circleci: lint_python_and_config"

File: gallery/plot_transforms.py [hunk 6]
```diff
@@ -119,6 +119,15 @@ def plot(imgs, with_orig=True, row_title=None, **imshow_kwargs):
...
 plot(blurred_imgs)
 
+####################################
+# GaussianNoise
+# ~~~~~~~~~~~~~
+# The :class:`~torchvision.transforms.GaussianNoise` transform
+# perturbs the input image with gaussian noise.
+noisy = T.GaussianNoise(mean=0.0, sigma=(0.1, 5))
+noisy_imgs = [noisy(orig_img) for _ in range(4)]
+plot(noisy_imgs)
```
Comment (vfdev-5) [inline gallery/plot_transforms.py:129]: This gives the following result: https://output.circle-artifacts.com/output/job/34f9c3d8-231e-4748-81b9-cda1724a9f16/artifacts/0/docs/auto_examples/plot_transforms.html#gaussiannoise which does not make sense. Let's reduce the number of transformed images and propose acceptable range of sigmas

File: test/test_transforms.py (hunk 3 above)
Comment (vfdev-5) [reply test/test_transforms.py:1321]: @parth-shastri to check the output type, please do `isinstance(out, PIL.Image.Image)`

File: gallery/plot_transforms.py (hunk 6 above)
Comment (parth-shastri) [reply gallery/plot_transforms.py]: I think we should not keep an accepted range of sigmas, just let the user decide how much distortion does he want ?

Comment (parth-shastri) [reply gallery/plot_transforms.py]: Fixed the documentation images in the documentation.

File: test/test_transforms.py (hunk 3 above)
Comment (vfdev-5) [reply test/test_transforms.py:1321]: @parth-shastri please address this comment

File: torchvision/transforms/transforms.py (hunk 1 above)
Comment (vfdev-5) [reply torchvision/transforms/transforms.py:1858]: @parth-shastri please address this comment

File: torchvision/transforms/transforms.py (hunk 2 above)
Comment (vfdev-5) [reply torchvision/transforms/transforms.py:1891]: @parth-shastri please address this comment

Comment (parth-shastri) [reply torchvision/transforms/transforms.py:1891]: Yes working on that

**Comment by parth-shastri** [comment]
@vfdev-5 can you please help me out and tell me why the tests are failing, are there any common causes of the failures?

File: test/test_transforms.py [hunk 7]
```diff
@@ -1314,6 +1314,25 @@ def test_gaussian_blur_asserts():
         transforms.GaussianBlur(3, "sigma_string")
 
 
+def test_gaussian_noise():
+    np_img = np.ones((100, 100, 3), dtype=np.uint8) * 255
+    img = F.to_pil_image(np_img, "RGB")
+    out = transforms.GaussianNoise(2.0, (0.1, 2.0))(img)
+    assert isinstance(out, PIL.Image.Image)
```
Comment (oke-aditya) [inline test/test_transforms.py:1321]: Probably tests are failing due to this. 
Also can you plz merge the latest main branch? That will avoid a few CI errors

Comment (parth-shastri) [reply test/test_transforms.py:1321]: Yes will do thanks!

**Comment by oke-aditya** [comment]
Are we stuck on this? @parth-shastri do you need help?

**Comment by parth-shastri** [comment]
> Are we stuck on this? @parth-shastri do you need help?
 Hey, thanks I am not able to decide on a valid sigma to display
and also is there a way that we can visualize the results in the documentation before the commit?
apart from the deciding of sigma is there anything else that I'm supposed to do?

**Comment by oke-aditya** [comment]
You can visualise the results by building the docs. 
Also I think you need to refactor the code to create a functional API. 

You can just try doing a grid search (or just try a few values and post the images here)

cc @vfdev-5 any suggestions for Sigma ?

**Comment by parth-shastri** [comment]
> 

Hi, tried out some sigmas and found that the range [0.1, 0.5] works.
gaussian_astronaut
gaussian_astronaut

File: torchvision/transforms/functional_tensor.py [hunk 8]
```diff
@@ -758,6 +756,29 @@ def gaussian_blur(img: Tensor, kernel_size: List[int], sigma: List[float]) -> Te
...
+        raise TypeError(f"img should be Tensor. Got {type(img)}")
+
+    _assert_image_tensor(img)
+    dtype = img.dtype if torch.is_floating_point(img) else torch.float32
+    img, need_cast, need_squeeze, out_dtype = _cast_squeeze_in(
+        img,
+        [
+            dtype,
+        ],
+    )
```
Comment (vfdev-5) [inline torchvision/transforms/functional_tensor.py:770]:
```suggestion
    img, need_cast, need_squeeze, out_dtype = _cast_squeeze_in(img, [dtype])
```

File: torchvision/transforms/functional_tensor.py [hunk 9]
```diff
@@ -758,6 +756,29 @@ def gaussian_blur(img: Tensor, kernel_size: List[int], sigma: List[float]) -> Te
...
+    # add the gaussian noise with the given mean and sigma.
+    normalize_img = img / 255.0
```
Comment (vfdev-5) [inline torchvision/transforms/functional_tensor.py:772]: We can't make this assumption that `img` range is [0, 255]. 
By default, we assume: [0, 1] for float images and [0, 255] for uint8 RGB images.
Let's not rescale image range and let user pick appropriate `mean` and `sigma` according to their data range.

File: torchvision/transforms/functional_tensor.py [hunk 10]
```diff
@@ -748,8 +748,6 @@ def gaussian_blur(img: Tensor, kernel_size: List[int], sigma: List[float]) -> Te
             kernel.dtype,
         ],
     )
-
-    # padding = (left, right, top, bottom)
```
Comment (vfdev-5) [inline torchvision/transforms/functional_tensor.py:752]: Let's keep this comment

File: torchvision/transforms/functional.py [hunk 11]
```diff
@@ -1404,6 +1404,55 @@ def gaussian_blur(img: Tensor, kernel_size: List[int], sigma: Optional[List[floa
...
+        PIL Image or Tensor: Gaussian Blurred version of the image.
+    """
+    if not torch.jit.is_scripting() and not torch.jit.is_tracing():
+        _log_api_usage_once(gaussian_noise)
+
+    if sigma is None:
+        raise ValueError("The value of sigma cannot be None.")
+
+    if sigma is not None and not isinstance(sigma, (int, float, list, tuple)):
+        raise TypeError(f"sigma should be either float or sequence of floats. Got {type(sigma)}")
```
Comment (vfdev-5) [inline torchvision/transforms/functional.py:1433]: Something is unclear here about whether sigma is `Optional` or not. Docstring and typehint says optional, but you raise an error if sigma is None. Let's remove optional and assume that sigma is not None

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the update @parth-shastri
I left few other comment to improve the PR

File: torchvision/transforms/functional.py (hunk 11 above)
Comment (vfdev-5) [reply torchvision/transforms/functional.py:1403]: @parth-shastri check above comment and the code. Why do you check `sigma is None` if sigma is not intended to be `Optional`.

File: torchvision/transforms/functional.py [hunk 12]
```diff
@@ -1375,6 +1375,54 @@ def gaussian_blur(img: Tensor, kernel_size: List[int], sigma: Optional[List[floa
...
+    if isinstance(sigma, (list, tuple)) and len(sigma) == 1:
+        sigma = [sigma[0], sigma[0]]
+    if isinstance(sigma, (list, tuple)) and len(sigma) == 2:
+        sigma = torch.empty(1).uniform_(sigma[0], sigma[1]).item()
```
Comment (vfdev-5) [inline torchvision/transforms/functional.py:1407]: Functional code should not perform any random calls, this should be done in Transforms

Comment (parth-shastri) [reply torchvision/transforms/functional.py:1407]: Yes will resolve that!

Comment (parth-shastri) [reply torchvision/transforms/functional.py:1407]: yes will resolve this

Comment (parth-shastri) [reply torchvision/transforms/functional.py]: pushed,

**Comment by parth-shastri** [comment]
@vfdev-5 is there anything more to do?

**Comment by parth-shastri** [comment]
@oke-aditya @vfdev-5 can you please highlight any changes I need to make in order to merge this pull request?

**Comment by javirk** [comment]
Hi. I think it would be valuable to have Gaussian Noise as a transformation. What is the state of this at the moment? Should I take over this PR in order to finally merge it?

**Comment by vfdev-5** [comment]
The problem with this PR right now is that it should add code compatible to the new v2 API.
I think it would be easier to start from a new PR and add new transform and all functional part. @javirk if you are interested in working on that, feel free to provide another PR. You may cherry-pick some commits on the gallery to acknowledge the work of @parth-shastri.

**Comment by parth-shastri** [comment]
> The problem with this PR right now is that it should add code compatible to the new v2 API. I think it would be easier to start from a new PR and add new transform and all functional part. @javirk if you are interested in working on that, feel free to provide another PR. You may cherry-pick some commits on the gallery to acknowledge the work of @parth-shastri.


Hi I can take this up, do you want me to create another pull request compatible with the new v2 API or shall I continue here?

**Comment by NicolasHug** [comment]
Thanks a lot for your PR @parth-shastri . We just merged https://github.com/pytorch/vision/pull/8381 which adds gaussian noise, but thanks again for the original implementation!

**Pull request closed without merging**